### PR TITLE
Keyboard fixes

### LINKED ISSUE #7047: Auto hide tabs when keyboard is open
When the keyboard is opened in a tabs page (position=bottom), the tabs appear on top of the keyboard.

The expected behavior is to hide the tabs when the keyboard is visible

**Steps to reproduce:**
1. Create a starter tabs template 
2. Include an ion-input in the page
3. Run it on a device and focus on the input element

Maybe a CSS class can be added to the page when the keyboard is open (.keyboard-is-open) so that the behavior can be controlled

This issue is for Ionic 2

Cordova CLI: 5.2.0
Gulp version:  CLI version 3.9.0
Gulp local:   Local version 3.9.1
Ionic Framework Version: 2.0.0-beta.9
Ionic CLI Version: 2.0.0-beta.30
Ionic App Lib Version: 2.0.0-beta.16
OS:
Node Version: v4.3.2

### LINKED ISSUE #9699: Keyboard TAB key traversal of input fields
**Ionic version:**  (check one with "x")
[ ] **1.x**
[x] **2.x**

**I'm submitting a ...**  (check one with "x")
[x] bug report
[ ] feature request
[ ] support request => Please do not submit support requests here, use one of these channels: https://forum.ionicframework.com/ or http://ionicworldwide.herokuapp.com/

**Current behavior:**
With external keyboard on iOS (or using keyboard in iOS emulator), i'm not able to traverse inputs in the form with TAB key. For a moment it switches the focus to the next field, but then the focus is lost. This is happening in default Ionic2 app created from scratch. However, when I add plugin `<plugin name="cordova-plugin-wkwebview-engine" spec="https://github.com/driftyco/cordova-plugin-wkwebview-engine.git"/>
  `, TAB traversal works correctly and as expected.

The issue is there on RC 4 and also in nightly.

**Expected behavior:**
Tab traversal should be working in a default setup without any special plugin.

**Steps to reproduce:**
out

1. `ionic start testapp --v2`
2. create a simple login form (taken from ionic-conference-app)
3. run iOS emulator
4. make username field focused
5. press TAB to switch to password field

the focus is lost.
Reproducible only in emulator, not on web.

http://plnkr.co/edit/xfTexJHLaW93MPAeHH6K?p=preview

**Other information:**

**Ionic info:** (run `ionic info` from a terminal/cmd prompt and paste output below):

```
insert the output from ionic info here
```
Cordova CLI: 6.3.1
Ionic Framework Version: 2.0.0-rc.4
Ionic CLI Version: 2.1.17
Ionic App Lib Version: 2.1.7
Ionic App Scripts Version: 0.0.47
ios-deploy version: 1.9.0
ios-sim version: 5.0.13
OS: macOS Sierra
Node Version: v6.3.1
Xcode version: Xcode 8.2 Build version 8C38

### LINKED ISSUE #10183: [BUG] content disappears when keyboard is up
**Ionic version:**  (check one with "x")
[ ] **1.x**
[x] **2.x**

**I'm submitting a ...**  (check one with "x")
[x] bug report
[ ] feature request
[ ] support request => Please do not submit support requests here, use one of these channels: https://forum.ionicframework.com/ or http://ionicworldwide.herokuapp.com/

**Current behavior:**
When having slides, with a scroll, as suggested [here](https://github.com/driftyco/ionic/issues/9839), in the last comment, the content disappears when opening the keyboard, except for the first input.

Note: this happens on device, in this example I am using an android emulator, but it also happens if you `serve` and open the page via iPhone.

[Here ](https://www.youtube.com/watch?v=C9w15LK10w0) is a video demonstration

**Expected behavior:**
Normal input behavior

**Steps to reproduce:**
As said, I am running this on a device, so plunkr won't work.
This is the relevant code:
```
<ion-slides #slider pager class="swiper-no-swiping">
    <ion-slide class="scroll-content">
        <ion-scroll scrollY="true">
            <h2>Tab title</h2>
            <ion-list>
                <ion-item *ngFor="let i of [1,2,3,4,5,6,7,8,9,10]">
                    <ion-label floating>Text field</ion-label>
                    <ion-input type="text" [(ngModel)]="text"></ion-input>
                </ion-item>
            </ion-list>
        </ion-scroll>
    </ion-slide>
</ion-slides>
```

This is the CSS:
```
ion-scroll {
	position: absolute;
	top: 0;
	bottom: 0;
	width: 100%;
}
```

**Ionic info:** (run `ionic info` from a terminal/cmd prompt and paste output below):

```
Your system information:

 ordova CLI: 6.4.0
Ionic Framework Version: 2.0.0
Ionic CLI Version: 2.2.1
Ionic App Lib Version: 2.2.0
Ionic App Scripts Version: 0.0.47
ios-deploy version: Not installed
ios-sim version: Not installed
OS: Windows 10
Node Version: v6.9.1
Xcode version: Not installed
```

### LINKED ISSUE #10187: issue(prompt alert): Should show in middle of the screen as per native alert
**Ionic version:**  (check one with "x")
[ ] **1.x**
[x] **2.x**

**I'm submitting a ...**  (check one with "x")
[ ] bug report
[x] issue
[ ] feature request
[ ] support request => Please do not submit support requests here, use one of these channels: https://forum.ionicframework.com/ or http://ionicworldwide.herokuapp.com/

**Current behavior:**
A prompt alert on a large device like an iPad (on iOS) is being displayed at the top half of the screen:

image

I am assuming this is done because then the keyboard appears, the alert will be in the right place.

**Expected behavior:**
This needs to work more like a native app. The prompt to appear in the center of the screen. When the software keyboard opens up, the alert needs to slide into the correct position near the top of the screen.

**Steps to reproduce:**
Create any [prompt alert](http://ionicframework.com/docs/v2/components/#alert-prompt) on an iPad and note it's positioning.  

**Ionic info:** (run `ionic info` from a terminal/cmd prompt and paste output below):

```
Cordova CLI: 6.5.0
Ionic Framework Version: 2.0.0
Ionic CLI Version: 2.2.1
Ionic App Lib Version: 2.2.0
Ionic App Scripts Version: 1.0.0
ios-deploy version: 1.9.0
ios-sim version: 5.0.10
OS: macOS Sierra
Node Version: v6.9.1
Xcode version: Xcode 8.2.1 Build version 8C1002
```

### LINKED ISSUE #10393: clearInput doesn't seem to work on IOS
**Ionic version:**  (check one with "x")
[ ] **1.x**
[x] **2.x**

**I'm submitting a ...**  (check one with "x")
[x] bug report
[ ] feature request
[ ] support request => Please do not submit support requests here, use one of these channels: https://forum.ionicframework.com/ or http://ionicworldwide.herokuapp.com/

**Current behavior:**
I want to add clearInput attribute on my ion-inputs. My ion-inputs have ngModel attributes and some have keyup functions bind to them. On IOS, the clearInput icon shows up on focus but nothing happends when clicking it besides the loss of focus on the ion-input. It does work on Android.

**Expected behavior:**
I expect inputs to be empty after clicking on the clearInput icon.

**Steps to reproduce:**

**Related code:**

```
<ion-item>
	<ion-input [(ngModel)]="lastName" clearInput></ion-input>
</ion-item>
```

**Other information:**

**Ionic info:** (run `ionic info` from a terminal/cmd prompt and paste output below):

```
Cordova CLI: 6.4.0
Ionic Framework Version: 2.0.1
Ionic CLI Version: 2.2.1
Ionic App Lib Version: 2.2.0
Ionic App Scripts Version: 1.0.0
ios-deploy version: 1.9.1
ios-sim version: Not installed
OS: OS X El Capitan
Node Version: v7.0.0
Xcode version: Xcode 8.2 Build version 8C38
```

### LINKED ISSUE #10552: Tabs being pushed up by keyboard when searchbar is blurred (iOS 10.2)
**Ionic version:**  (check one with "x")
[ ] **1.x**
[x] **2.x**

**I'm submitting a ...**  (check one with "x")
[x] bug report
[ ] feature request
[ ] support request => Please do not submit support requests here, use one of these channels: https://forum.ionicframework.com/ or http://ionicworldwide.herokuapp.com/

**Current behavior:**
In iOS 10.2, I am having an issue with my tabs when blurring a searchbar component. When the keyboard slides down, the tabs suddenly appear in the middle of the screen for about 300-500ms, and then drop back down to the bottom.

**Expected behavior:**
I would expect that the tabs would remain hidden behind the keyboard, and when it slides back down, they would be revealed again, in the same position.

**Steps to reproduce:**
This only appears when you're actually previewing the app on an iOS device; not in the browser (since, obviously, the soft keyboard is not shown in browser).

**Ionic info:** (run `ionic info` from a terminal/cmd prompt and paste output below):

```
Cordova CLI: 6.5.0 
Ionic Framework Version: 2.0.1
Ionic CLI Version: 2.2.1
Ionic App Lib Version: 2.2.0
Ionic App Scripts Version: 1.1.0
ios-deploy version: Not installed
ios-sim version: 5.0.13 
OS: macOS Sierra
Node Version: v6.9.4
Xcode version: Xcode 8.2.1 Build version 8C1002
```

### LINKED ISSUE #11251: Content Does Not Resize on Keyboard Open
**Ionic version:**  (check one with "x")
[ ] **1.x**
[x] **2.x**
[ ] **3.x**

**I'm submitting a ...**  (check one with "x")
[x] bug report
[ ] feature request
[ ] support request => Please do not submit support requests here, use one of these channels: https://forum.ionicframework.com/ or http://ionicworldwide.herokuapp.com/

**Current behavior:**
When opening keyboard in iOS, content does not resize. Content does resize when closing keyboard, which causes blank space at the top or bottom of screen, followed by a second resize that corrects the screen size once the keyboard is fully closed.

**Expected behavior:**
Content should resize with opening keyboard and both footer and header should remain visible so long as `this.keyboard.disableScroll(true);` is called.

**Other information:**
Video of issue: https://streamable.com/4tkoj

**Ionic info:** (run `ionic info` from a terminal/cmd prompt and paste output below):

```
Cordova CLI: 6.5.0 
Ionic Framework Version: 3.0.1
Ionic CLI Version: 2.2.2
Ionic App Lib Version: 2.2.1
Ionic App Scripts Version: 1.3.0
ios-deploy version: 1.9.0 
ios-sim version: 5.0.13 
OS: macOS Sierra
Node Version: v7.3.0
Xcode version: Xcode 8.3.1 Build version 8E1000a
```

### LINKED ISSUE #11291: Searchbar - Keyboard close will push tabs up.
**Ionic version:**  (check one with "x")
[ ] **1.x**
[ ] **2.x**
[x] **3.x**

**I'm submitting a ...**  (check one with "x")
[x] bug report
[ ] feature request
[ ] support request => Please do not submit support requests here, use one of these channels: https://forum.ionicframework.com/ or http://ionicworldwide.herokuapp.com/

**Current behavior:**

If you click outside the keyboard then tabs will push up. 

searchbar

**Expected behavior:**

Same as the close behavior via KeyboardAccessoryBar

**Steps to reproduce:**

**Related code:**

```
    <ion-searchbar placeholder="Zoeken op kenteken/titel/bj/notitie" (ionClear)="getItems($event)" (input)="getItems($event)"></ion-searchbar>

```

**Other information:**

**Ionic info:** (run `ionic info` from a terminal/cmd prompt and paste output below):

```
insert the output from ionic info here
```
 Your system information:

    Cordova CLI: 6.5.0 
    Ionic Framework Version: 3.0.1
    Ionic CLI Version: 3.0.0-beta7
    ios-deploy version: 1.9.0 
    ios-sim version: 5.0.13 
    OS: macOS Sierra
    Node Version: v6.9.4
    Xcode version: Xcode 8.3.1 Build version 8E1000a

### LINKED ISSUE #11325: Keyboard overlays popover when using forms
**Ionic version:**  (check one with "x")
[ ] **1.x**
[ ] **2.x**
[x] **3.x**

**I'm submitting a ...**  (check one with "x")
[x] bug report
[ ] feature request
[ ] support request => Please do not submit support requests here, use one of these channels: https://forum.ionicframework.com/ or http://ionicworldwide.herokuapp.com/

**Current behavior:**
So I need to add a small form inside a popover. It's a "add to wishlist" popover. So the user either selects an existing wishlist or creates a new wishlist and add the item at the same time. When I tried that in the device, I got a weird result where the keyboard was overlaying the popover and the form couldn't be seen.

Image: http://imgur.com/a/payjQ

**Expected behavior:**
I'm not sure if that is somewhat achievable, but it would be nice if there was a scroll, just like in a normal page with input forms.

**Steps to reproduce:**
Create the popover component and try it out *on a device*

**Related code:**

.html
```
<ion-content>
	<ion-list>
		<ion-list-header>Wishlists</ion-list-header>
		<button ion-item *ngFor="let wishlist of wishlists" (click)="addToWishList(wishlist.id)">{{wishlist.name}}</button>
	</ion-list>
	<div *ngIf="!showInput">
		<button ion-item detail-none (click)="create()" class="pink add-button">Create <ion-icon name="add" class="right"></ion-icon></button>
	</div>
	<div *ngIf="showInput">
		<ion-item>
			<ion-input placeholder="Name.." [(ngModel)]="wishlistName" maxlength="25"></ion-input>
		</ion-item>
		<button ion-button clear (click)="createNewWishlist()" class="right">Add</button>
	</div>
</ion-content>
```

.ts
```
import { ViewController } from 'ionic-angular';
import { Component } from '@angular/core';
@Component({
  selector: 'wishlists-popover',
  templateUrl: 'wishlists-popover.html',
})
export class WishlistsPopover {
  showInput = false;
  wishlistName = "";
  wishlists = [
    {
      id: 1,
      name: "Dream's adventure"
    },
    {
      id: 2,
      name: "Weekend",
    },
    {
      id: 3,
      name: "Family Time"
    },
  ];

  constructor(public viewCtrl: ViewController) { }

  create() {
    this.showInput = true;
  }
  createNewWishlist() {
    this.wishlists.push({ id: 5, name: this.wishlistName });
    this.showInput = false;
    this.wishlistName = "";
  }

  addToWishList(id) {
    console.log("added to " + id);
  }

  close() {
    this.viewCtrl.dismiss();
  }
}
```

 
```
Cordova CLI: 6.3.1
Ionic Framework Version: 3.0.1
Ionic CLI Version: 2.2.1
Ionic App Lib Version: 2.2.0
Ionic App Scripts Version: 1.3.0
ios-deploy version: Not installed
ios-sim version: Not installed
OS: Linux 4.4
Node Version: v6.9.1
Xcode version: Not installed
```

### LINKED ISSUE #11389: ion-input becomes unresponsive after losing focus
**Ionic version:**  (check one with "x")
[ ] **1.x**
[x ] **2.x**
[ ] **3.x**

**I'm submitting a ...**  (check one with "x")
[x ] bug report
[ ] feature request
[ ] support request => Please do not submit support requests here, use one of these channels: https://forum.ionicframework.com/ or http://ionicworldwide.herokuapp.com/

**Current behavior:**
1. Click on ion-input and type something.
2. To close the keyboard click outside the keyboard and the input box. 
3. During the time the keyboard is closing, try to scroll up or down.
4. When you try to edit the textbox, its unresponsive.
 
This happens inconsistently, I need to try the steps 2 and 3 around 10-15 times for the issue to happen.

**Expected behavior:**
The textbox should be editable at all times, shouldn't become unresponsive.

**Steps to reproduce:**
I created the ionic 2.x tabs starter project and added the following html in the about page. 
I am able to replicate this on iPad (iOS version 10.2.1).
This issue is happening only on ion-input and not on input textbox.

Again as mentioned above, you need to try the steps 2 and 3 around 10-15 times to see the issue. On few occasions, I am able to reproduce it in 2-3 attempts.

**Related code:**

```
<ion-content padding>
  <ion-list>
    <ion-item><div>Some text</div></ion-item>
    <ion-item><div>Some text</div></ion-item>
    <ion-item><div>Some text</div></ion-item>
    <ion-item><div>Some text</div></ion-item>
    <ion-item><div>Some text</div></ion-item>
    <ion-item><div>Some text</div></ion-item>
    <ion-item><input type="text" placeholder="normal input"></ion-item>
    <ion-item><ion-input type="text" placeholder="ion input"></ion-input></ion-item>
  </ion-list>
</ion-content>
```

**Other information:**
I looked into the ion-input code and found that ionic creates a cloned input element during the time user focuses on the textbox and its supposed to be deleted once the keyboard is opened up.
I have noticed that when this issue happens, the cloned input element is not deleted for some reason.
Here's the screenshot showing the cloned input element present:
screen shot 2017-04-26 at 6 58 52 pm

Since the cloned input element is not deleted, the user cannot edit the field and any changes to the model is also not shown on the UI to the user.

**Ionic info:** (run `ionic info` from a terminal/cmd prompt and paste output below):

```
Cordova CLI: 6.0.0 
Ionic CLI Version: 2.1.18
Ionic App Lib Version: 2.1.9
ios-deploy version: 1.9.1 
ios-sim version: 5.0.13 
OS: macOS Sierra
Node Version: v6.1.0
Xcode version: Xcode 8.2.1 Build version 8C1002
```

### LINKED ISSUE #11484: ion-searchbar focus issues
**Ionic version:**  (check one with "x")
[ ] **1.x**
[ ] **2.x**
[X ] **3.x**

**I'm submitting a ...**  (check one with "x")
[X] bug report
[ ] feature request
[ ] support request => Please do not submit support requests here, use one of these channels: https://forum.ionicframework.com/ or http://ionicworldwide.herokuapp.com/

**Current behavior:**
ion-searchbar has focus issues. Does not lose focus when touching other areas of the screen. Incidentally the keyboard stays up as well and persists through screen changes, effectively breaking the app.

https://www.youtube.com/watch?v=kXSUjTJPRqM   <- watch

**Expected behavior:**
The search bar should lose focus when other areas of the screen are touched. Further, the search bar should lose focus when transitioning screens.

**Other information:**
https://www.youtube.com/watch?v=kXSUjTJPRqM

**Ionic info:** (run `ionic info` from a terminal/cmd prompt and paste output below):
Cordova CLI: 6.4.0
Ionic Framework Version: 3.1.0
Ionic CLI Version: 2.1.14
Ionic App Lib Version: 2.1.7
Ionic App Scripts Version: 1.3.4
ios-deploy version: Not installed
ios-sim version: Not installed
OS: Windows 10
Node Version: v4.4.1
Xcode version: Not installed

### LINKED ISSUE #11566: ion-searchbar bug on 3.1.0, 3.1.1, 3.2.0 for non-latin languages.
**Ionic version:**  (check one with "x")
[ ] **1.x**
[ ] **2.x**
[ X] **3.x**

**I'm submitting a ...**  (check one with "x")
[X ] bug report
[ ] feature request
[ ] support request => Please do not submit support requests here, use one of these channels: https://forum.ionicframework.com/ or http://ionicworldwide.herokuapp.com/

**Current behavior:**

Replication environment:  Mac OS X (for PWA)
Different behaviours in Chrome  58.0.3029.96  and Safari Version 10.1 (12603.1.30.0.34) (see videos)

Currently I have replicated the bug on the ionic-preview-app on ionic versions 3.1.1, 3.1.0, 3.2.0

The bug does not happen on 3.0.1.

Description:  The key up on the ion-searchbar seems to be broken since 3.1.0 onwards for non latin keyboards.  For example I am using a chinese Hanyu pin yin keyboard to type "青山"  (qingshan).

However on Safari, the ion-searchbar goes like "qqqiqiqingqinqingqing sqing sqing shqing... ".  See video.
On Chrome, the ion-searchbar seems to dismiss the language bar intended to popup.

**Expected behavior:**

The ion-searchbar should work normally with the different language inputs.

See video.

**Steps to reproduce:**

1.  check out and build the ionic-preview-app and build 
2. ionic serve
3. scroll to the ion-searchbar page.
4. change input to chinese and type "qingshan".  notice on Chrome the input for chinese does not appear, and on Safari multiple letters appear.

5.  Change package.json ionic-angular dependency to 3.0.1
6.  remove node_modules and npm install again
7. notice that things work normally.

**Related code:**

```
insert any relevant code here
```
https://github.com/driftyco/ionic-preview-app

**Other information:**

**Ionic info:** (run `ionic info` from a terminal/cmd prompt and paste output below):

```
insert the output from ionic info here
```
Cordova CLI: 6.5.0 
Ionic Framework Version: 3.0.1  (or 3.1.1 for the bug)
Ionic CLI Version: 2.2.2
Ionic App Lib Version: 2.2.1
Ionic App Scripts Version: 1.3.5
ios-deploy version: 1.9.1 
ios-sim version: 5.0.2 
OS: macOS Sierra
Node Version: v7.9.0

bug on safari
Bug on ion-searchbar on 3.1.0 and 3.1.1

expected behaviour
Expected behaviour

### PR DESCRIPTION
**Pull request by manucorporat**
DO NOT MERGE

fixes #7047
fixes #9699
fixes #11566
fixes #11484
fixes #11389
fixes  #11325
fixes #11291
fixes #11251 ?
fixes #10552
fixes #10393
fixes #10183
fixes #10187

### CODE REVIEW

File: src/components/datetime/datetime.ts [hunk 1]
```diff
@@ -259,6 +259,7 @@ import { dateValueRange, renderDateTime, renderTextFormat, convertDataToISO, con
             'ion-button="item-cover" ' +
             '[attr.aria-labelledby]="_labelId" ' +
             '[attr.aria-disabled]="_disabled" ' +
+            '(click)="_click($event)" ' +
```
Comment (manucorporat) [inline src/components/datetime/datetime.ts]: revert

File: src/components/input/input.scss [hunk 2]
```diff
@@ -119,7 +87,6 @@ input.text-input:-webkit-autofill {
   pointer-events: auto;
 }
 
-
```
Comment (manucorporat) [inline src/components/input/input.scss]: revert

File: src/components/item/item-sliding.ts [hunk 3]
```diff
@@ -346,6 +346,7 @@ export class ItemSliding {
     }
     if (openAmount === 0) {
       this._tmr = platform.timeout(() => {
+        // change in the code
```
Comment (manucorporat) [inline src/components/item/item-sliding.ts]: revert

File: src/components/radio/radio-button.ts [hunk 4]
```diff
@@ -134,6 +134,8 @@ export class RadioButton extends Ion implements IonicTapInput, OnDestroy, OnInit
     }
   }
 
+  initFocus() { }
```
Comment (manucorporat) [inline src/components/radio/radio-button.ts]: revert

File: src/components/select/select.ts [hunk 5]
```diff
@@ -206,10 +206,6 @@ export class Select extends BaseInput<any> implements OnDestroy {
 
   @HostListener('click', ['$event'])
   _click(ev: UIEvent) {
-    if (ev.detail === 0) {
```
Comment (manucorporat) [inline src/components/select/select.ts]: review, ask @adamdbradley

File: src/components/tabs/tabs.ts [hunk 6]
```diff
@@ -261,10 +266,34 @@ export class Tabs extends Ion implements AfterViewInit, RootNode, ITabs {
       viewCtrl._setContent(this);
       viewCtrl._setContentRef(elementRef);
     }
+
+    if (keyboard) {
```
Comment (manucorporat) [inline src/components/tabs/tabs.ts]: if the webview is not resizes this code does not really matter, read config.keyboardResizes?

File: src/platform/keyboard.ts [hunk 7]
```diff
@@ -26,14 +26,64 @@ import { Platform } from './platform';
...
+    config: Config,
+    private _plt: Platform,
+    private _zone: NgZone,
+    private _dom: DomController
+  ) {
     this.focusOutline(config.get('focusOutline'));
 
-    const win = _plt.win();
+    const win = <any>_plt.win();
+    // const usingIonicEngine = win.Ionic && win.Ionic.KeyboardEvents === true;
```
Comment (manucorporat) [inline src/platform/keyboard.ts]: think back-compatible support

File: src/platform/keyboard.ts [hunk 8]
```diff
@@ -26,14 +26,64 @@ import { Platform } from './platform';
...
+      this._zone.run(() => {
+        this.willShow.emit();
+      });
+    }, { zone: false, passive: true });
+
+    this._plt.registerListener(win, 'keyboardWillHide', () => {
+      this._zone.run(() => {
+        this.willHide.emit();
+      });
+      // this._plt.focusOutActiveElement();
```
Comment (manucorporat) [inline src/platform/keyboard.ts]: review this, ask @jgw96 @adamdbradley @mhartington

File: src/platform/keyboard.ts [hunk 9]
```diff
@@ -158,7 +208,7 @@ export class Keyboard {
     let isKeyInputEnabled = false;
     let unRegMouse: Function;
     let unRegTouch: Function;
-    const evOpts = { passive: true, zone: false };
+    const evOpts = { passive: true, zone: false, capture: true };
```
Comment (manucorporat) [inline src/platform/keyboard.ts]: why did I change this? probably revert or add comment explaining why

File: src/tap-click/tap-click.ts [hunk 10]
```diff
@@ -114,7 +116,7 @@ export class TapClick {
     this.startCoord = null;
   }
 
-  pointerCancel(ev: UIEvent) {
+  pointerCancel(ev: any) {
```
Comment (manucorporat) [inline src/tap-click/tap-click.ts]: revert

File: src/util/base-input.ts [hunk 11]
```diff
@@ -210,12 +209,17 @@ export class BaseInput<T> extends Ion implements CommonInput<T> {
    */
   _fireFocus() {
     if (this._isFocus) {
+      console.error('BaseInput already focused');
```
Comment (manucorporat) [inline src/util/base-input.ts]: remove this before merging!!!!!!!!

File: src/util/base-input.ts [hunk 12]
```diff
@@ -255,9 +262,20 @@ export class BaseInput<T> extends Ion implements CommonInput<T> {
...
+    if (isArray(val) || isString(val)) {
+      return val.length > 0;
+    }
+    return true;
+  }
+
+  /**
+   * @hidden
+   */
+  focusNext() {
```
Comment (manucorporat) [inline src/util/base-input.ts]: where is focusNext() used? need to review it
how does it interact between input, toggle, checkbox (look at items/inputs)

File: src/util/form.ts [hunk 13]
```diff
@@ -7,32 +7,38 @@ import { removeArrayItem } from './util';
...
    */
-  tabFocus(currentInput: any) {
-    let index = this._inputs.indexOf(currentInput);
-    if (index > -1 && (index + 1) < this._inputs.length) {
-      let nextInput = this._inputs[index + 1];
+  tabFocus(currentInput: IonicFormInput) {
+    const inputs = this._inputs;
+    let index = this._inputs.indexOf(currentInput) + 1;
+    if (index > 0 && index < inputs.length) {
+      let nextInput = inputs[index];
```
Comment (manucorporat) [inline src/util/form.ts]: use var

Comment (manucorporat) [inline src/util/form.ts]: add unit tests! make sure I didn't break anything

File: src/util/scroll-view.ts [hunk 14]
```diff
@@ -68,6 +68,20 @@ export class ScrollView {
     this._eventsEnabled = true;
   }
 
+  setScrolling(isScrolling: boolean, ev: ScrollEvent) {
```
Comment (manucorporat) [inline src/util/scroll-view.ts]: review with @adamdbradley (explain use case)
add support for JS-scrolling? maybe not needed

add new unit tests

File: src/components/input/input.ts [hunk 15]
```diff
@@ -84,92 +84,44 @@ import { Platform } from '../../platform/platform';
...
+    '[max]="max" ' +
+    '[step]="step" ' +
+    '[placeholder]="placeholder" ' +
+    '[disabled]="disabled" ' +
+    '[readonly]="readonly">' +
+
+  '<textarea #textInput *ngIf="_isTextarea" class="text-input" ' +
+    '[(ngModel)]="value" ' +
+    '(blur)="onBlur()" ' +
+    '(focus)="onFocus()" ' +
```
Comment (manucorporat) [inline src/components/input/input.ts]: we could use non-zoned focusin/focusout programatically!!

File: src/components/input/input.ts [hunk 16]
```diff
@@ -84,92 +84,44 @@ import { Platform } from '../../platform/platform';
...
-
-    this._autoFocusAssist = config.get('autoFocusAssist', 'delay');
-    this._autoComplete = config.get('autocomplete', 'off');
-    this._autoCorrect = config.get('autocorrect', 'off');
-    this._keyboardHeight = config.getNumber('keyboardHeight');
-    this._useAssist = config.getBoolean('scrollAssist', false);
-    this._usePadding = config.getBoolean('scrollPadding', this._useAssist);
-
-    if (elementRef.nativeElement.tagName === 'ION-TEXTAREA') {
-      this._type = TEXTAREA;
```
Comment (manucorporat) [inline src/components/input/input.ts]: make sure type has the correct value.

add unit test

File: src/components/datetime/datetime.ts [hunk 17]
```diff
@@ -473,12 +475,7 @@ export class DateTime extends BaseInput<DateTimeData> implements AfterContentIni
     return convertDataToISO(this.value);
   }
 
-  @HostListener('click', ['$event'])
   _click(ev: UIEvent) {
-    // do not continue if the click event came from a form submit
```
Comment (manucorporat) [inline src/components/datetime/datetime.ts]: review, ask @adamdbradley

File: src/util/base-input.ts [hunk 18]
```diff
@@ -210,12 +209,17 @@ export class BaseInput<T> extends Ion implements CommonInput<T> {
...
       return;
     }
     assert(this._init, 'component was not initialized');
     assert(NgZone.isInAngularZone(), '_fireFocus: should be zoned');
 
+    // immediately set focus
     this._isFocus = true;
+    console.debug('BaseInput: focused:', this);
+    this._form && this._form.setAsFocused(this);
+    this._inputFocusChanged(true);
```
Comment (manucorporat) [inline src/util/base-input.ts]: we might not need `_inputFocusChanged`

**Comment by Khalid-Nowaf** [comment]
Hi @https://github.com/manucorporat, do you think this PR can fix #11745 ?

**Pull request merged**